cmd/kube-dns/SPEC.md

```diff
@@ -0,0 +1,184 @@
+- Question Example:
+  - `headless.default.svc.cluster.local. IN A`
+- Answer Example:
+```
+    headless.default.svc.cluster.local. 4 IN A 10.3.0.1
+    headless.default.svc.cluster.local. 4 IN A 10.3.0.2
+    headless.default.svc.cluster.local. 4 IN A 10.3.0.3
+```
+
+There must also be an `A` record of the following form for each _ready_
+endpoint with _hostname_ of `<hostname>` and IP address `<endpoint-ip>`.
+If there are multiple IP addresses for a given _hostname_, then there
+must be one such `A` record returned for each IP.
+- Record Format:
+  - `<hostname>.<service>.<ns>.svc.<zone>. <ttl> IN A <endpoint-ip>`
+- Question Example:
+  - `my-pet.headless.default.svc.cluster.local. IN A`
+- Answer Example:
+  - `my-pet.headless.default.svc.cluster.local. 4 IN A 10.3.0.100`
+
+#### `SRV` Records
+For each combination of _ready_ endpoint with _hostname_ of `<hostname>`, and
+port in the Service with name `<port>` and number `<port-number>` using
+protocol `<proto>`, an `SRV` record of the following form must exist.
+- Record Format:
+   - `_<port>._<proto>.<service>.<ns>.svc.<zone>. <ttl> IN SRV <weight> <priority> <port-number> <hostname>.<service>.<ns>.svc.<zone>.`
+
+This implies that if there are **N** _ready_ endpoints and the Service
+defines **M** named ports, there will be **N** :heavy_multiplication_x: **M**
+`SRV` RRs for the Service.
+
+The priority `<priority>` and weight `<weight>` are numbers as described
+in [RFC2782](https://tools.ietf.org/html/rfc2782) and whose values are not
+prescribed by this specification.
+
+Unnamed ports do not have an `SRV` record.
+
+- Question Example:
+  - `_https._tcp.headless.default.svc.cluster.local. IN SRV`
+- Answer Example:
+```
+        _https._tcp.headless.default.svc.cluster.local. 4 IN SRV 10 100 443 my-pet.headless.default.svc.cluster.local.
+        _https._tcp.headless.default.svc.cluster.local. 4 IN SRV 10 100 443 my-pet-2.headless.default.svc.cluster.local.
+        _https._tcp.headless.default.svc.cluster.local. 4 IN SRV 10 100 443 438934893.headless.default.svc.cluster.local.
+```
+
+The Additional section of the response may include the `A` records
+referred to in the `SRV` records.
+
+#### `PTR` Records
+
+Given a _ready_ endpoint with _hostname_ of `<hostname>` and IP address
+`<a>.<b>.<c>.<d>`, a `PTR` record of the following form must exist.
+- Record Format:
+  - `<d>.<c>.<b>.<a>.in-addr.arpa. <ttl> IN PTR <hostname>.<service>.<ns>.svc.<zone>.`
+- Question Example:
+  - `100.0.3.10.in-addr.arpa. IN PTR`
+- Answer Example:
+  - `100.0.3.10.in-addr.arpa. 14 IN PTR my-pet.headless.default.svc.cluster.local.`
+
+### Deprecated Records
+Kube-DNS versions prior to implementation of this specification also replied
+with an `A` record of the form below for any values of `<a>`, `<b>`, `<c>`, and `<d>` between 0 and 254:
+- Record Format:
+  - `<a>-<b>-<c>-<d>.<ns>.pod.<zone>. <ttl> IN A <a>.<b>.<c>.<d>`
+
+This behavior is deprecated and not required to satisfy this specification.
```

Comment:
We should announce an intent to deprecate it in favor of named headless services and take feedback.  All use cases I'm aware of are headless like things or software that requires reverse lookup of a record to match the local IP, which named headless should solve.